### ISSUE TITLE
Conflicting API docs for RenderObjects

### ISSUE THREAD

**Issue by matthew-carroll** [flutter/flutter#115525]
The following screenshot shows the current API docs for the `detach()` method in `AbstractNode`. The instructions appear to conflict. We're told "first call their inherited detach method, and then detach all their children", but we're also told "this method should end with a call to the inherited method, as in super.detach()". We're told that "super.detach()" should be the first thing called, but also the last thing called.

There may be similar issues with other docs. The line "Implementations of this method should end with a call to the inherited method, as in super.detach()." is repeated in a similar fashion for some other similar APIs. I suggest doing an audit.

![Screen Shot 2022-11-16 at 8 14 02 PM](https://user-images.githubusercontent.com/7259036/202353963-d5804537-4cc1-4b41-b0cf-6922ce2778bb.png)

**Comment by exaby73** [comment]
Hello @matthew-carroll. Thank you for raising this issue. Labeling for insights and visibility from the team

**Comment by gnprice** [comment]
I did a quick grep for "should end with". There are indeed several methods that have the same contradiction in their docs:
* RenderObject.detach (the original report — and also the one I just ran into myself, causing me to find this thread)
* Layer.detach
* AbstractNode.detach

The first two of those were copied recently from the latter, in the deprecation of AbstractNode:
* #128973 
* #128467

Then there are three methods that have the same "should end with a call to the inherited method" language but are fine, with no contradictions:
* State.deactivate
* State.dispose
* Element.deactivate

I also grepped for "should start with" (and "should begin with", which had no results). That found several methods that say that, and *redundantly* (but consistently) also have the "first call their inherited … method" language:
* RenderObject.attach
* Layer.attach
* AbstractNode.attach

as well as several methods on State and Element and elsewhere that have the "should start with" without the redundancy.

---

So the audit comes down to:
* The contradiction appears in those three `detach` methods (which were one method until very recently).
* The corresponding three `attach` methods have a redundancy.

That redundancy on `attach` is in itself a bit confusing for the reader. Both would be good to fix.

That leaves the question of which way around is the right one for `detach`. Typically one would expect to detach in the opposite order from attaching, so that'd mean ending with the `super.detach()` call (since the `attach` docs are consistent that `super.attach()` comes first).

Then looking at the Git history to see if that sheds any light… well, that text was added here:
* #76021

Before that, there was only one version on each of `attach` and `detach`, both saying to call super first. Seems like the author and reviewers of #76021 just didn't notice the existing text there.

And that "first call their inherited … method" text on both methods was added in 2017 in:
* #9456

There's no discussion in the docs or in #9456 of why `attach` and `detach` surprisingly call for the same order. So my guess would be that the version added later in #76021 is the right one. If it even matters, that is — there's plenty of render objects doing it each way, and it may be that the order doesn't matter in practice anyway.

I guess the next step is to send a PR, and any further discussion of what the right answer is can go in review there.

**Comment by gnprice** [comment]
> There's no discussion in the docs or in #9456 of why `attach` and `detach` surprisingly call for the same order. So my guess would be that the version added later in #76021 is the right one. If it even matters, that is — there's plenty of render objects doing it each way, and it may be that the order doesn't matter in practice anyway.

Ah, no. On looking closer at the implementation of `detach` in these base classes, there's an assertion that would fire if a child is detached before its parent calls `super.detach`. So in fact `super.detach` must be called before detaching any of the children, just as the older part of the text says.

And looking again at those implementations on various render objects, many of them don't start with `super.detach()`… but the things they do before that are stopping controllers and removing listeners, not detaching children.

Anyway, I'll reflect those findings in the PR I'll send shortly.
